proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1030 @@
+(The term **unit** refers to the intermediate language of units which
+mediates between the compiler and the package system; see the section `Units`_
+for more information.) Non-Backpack
+libraries do not have a module substitution (since they have no
+signatures to fill).  A unit identifier with no free module variables
+(see below) uniquely identifies a definite library.  We will use the
+metavariable ``P`` to represent unit identifiers.
+
+.. _Module:
+
+A **module identifier** is either a concrete module or a module
+variable.  A concrete module consists of a module name (the module
+being identified) and a unit identifier (the library it is a member of.)
+A module variable consists only of a module name, and specifies the
+name of an unfilled requirement.  Instantiation takes place by
+substituting module variables with concrete module identifiers;
+e.g., ``p[A=<A>]:B`` is instantiated to ``p[A=q:C]:B`` by applying
+the substitution ``<A>`` maps to ``q:C``.  We may speak of the
+**free module variables** of a module or unit identifier in the
+conventional sense.  Module variables are bound by the requirements
+of their defining library.  We will use the metavariable ``M`` to
+represent module identifiers, and ``m`` to represent module names.
+
+Module identifiers are used to ascribe unique names to Haskell entities:
+a Haskell entity ``n`` defined in a module with the module identity
+``M`` (notice that the unique name of a declaration in a module is
+not well-specified without extra information, since only module
+names are specified in syntax) is ascribed the unique name ``M.n``:
+two identifiers are only equal if their unique names are equal.
+
+.. _ModuleSubst:
+
+A **module substitution** is a mapping from module names to modules identifiers.
+Any module name occurs only once in a substitution, and a substitution
+is in canonical form if it is sorted by module name.  We will use
+the metavariable ``S`` to represent substitutions.
+
+Module substitutions can be applied to identifiers::
+
+    -- Substitution on UnitId
+    (p[S])⟦S'⟧   = p[S⟦S'⟧]
+
+    -- Substitution on Module
+    <m>⟦⟧        = <m>
+    <m>⟦m=M,  S⟧ = M
+    <m>⟦m'=M, S⟧ = <m>⟦S⟧    (m ≠ m')
+    (P:m)⟦S⟧     = P⟦S⟧:m
+
+    -- Substitution on ModuleSubst (NOT substitution composition)
+    (m=M, S')⟦S⟧ = m=M⟦S⟧, S'⟦S⟧
+
+.. _`generalized unit id`:
+
+A unit identifier can be **generalized** by replacing every mapping from
+``m=M`` in its module substitution with
+``m=<m>``.  This concept is useful when defining an
+index over both indefinite and definite libraries. Naively, we would
+have to use ``Either ComponentId UnitId`` as the key, since indefinite
+libraries are uniquely identified by component identifiers, while
+definite libraries identified by unit identifiers with no free module
+variables. Generalization lets us embed ``ComponentId`` in ``UnitId``,
+allowing us to use unit identifiers as the key uniformly across all
+types of libraries.
+
+Units
```

Comment:
Example: library with two holes; another one which partially instantiated the original one.